icicle/appUtils/ntt/ntt.cuh

```diff
@@ -314,17 +355,47 @@ void ntt_inplace_batch_template(
                                                       // then max to allow more concurrent blocks on SM
   const int logn_shmem = is_shared_mem_enabled ? int(log(2 * num_threads) / log(2))
                                                : 0; // TODO: shared memory support only for types <= 32 bytes
-
   if (inverse) {
     if (is_shared_mem_enabled)
       ntt_template_kernel_shared<<<num_blocks, num_threads, shared_mem, stream>>>(
         d_inout, 1 << logn_shmem, d_twiddles, n, total_tasks, 0, logn_shmem);
+    cudaError_t cudaError = cudaGetLastError();
+    if(cudaError != cudaSuccess) {
+      printf("\nCUDA error: %s\n", cudaGetErrorString(cudaError));       
+      exit(1);
+    }
 
-    for (int s = logn_shmem; s < logn; s++) // TODO: this loop also can be unrolled
+    // E* d_inout_original = new E[n];
+    // for (int i = 0; i < n; i++) {
+    //   d_inout_original[i] = d_inout[i];
+    // }
+    for (int s = logn_shmem; s < logn - 1; s++) // TODO: this loop also can be unrolled
     {
       ntt_template_kernel<E, S>
-        <<<num_blocks, num_threads, 0, stream>>>(d_inout, n, d_twiddles, n, total_tasks, s, false);
+        <<<num_blocks / 2, num_threads / 2, 0, stream>>>(d_inout, n / 2, d_twiddles, n / 2, total_tasks / 2, s, false);
+      ntt_template_kernel<E, S>
```

Comment:
As far as I understand, your idea here is to take all the stages of Cooley-Tukey algorithm except the very last one and somehow parallelise the two halves of each stage? I think correctness breaks here, one issue is that the total number of butterflies you launch in each kernel call is `num_blocks * num_threads / 4`, for two kernel calls this is `num_blocks * num_threads / 2` so two times fewer butterflies than we need to complete one stage of Cooley-Tukey. I think this two-fold reduction in the number of butterflies is the reason for the acceleration of your NTT as compared to the baseline.
I believe that splitting one kernel into two here is possible in principle; however this does not lead to more parallelism, but to less - two kernels get called one-by-one while one kernel call may exploit GPU resources as much as possible. So for small NTTs this approach of splitting kernels might result in worse runtimes whereas for large NTTs it seems to be basically the same when accounted for the correct number of butterflies.

Reply:
Thank you for your detailed review. This approach does indeed break correctness. Could you please give me more idea on how to split a kernel into two here to get more parallelism?  In my use case, I want a fast 2<sup>26</sup> size NTT on bn254. Thank you very much.

Reply:
Could you please write to us at [hi@ingonyama.com](mailto:hi@ingonyama.com) so we can discuss in more detail? :)